### ISSUE TITLE
dist.all_to_all_single, when input tensors of different shapes result in undefined output behavior

### ISSUE THREAD

**Issue by GTX960SmashAircraftCarrier** [pytorch/pytorch#158016]
### 🐛 Describe the bug

Device: 4*A100, Driver Version: 535.161.08, CUDA Version: 12.4

For rank0~2, input shape[4, 4]. For rank3, input shape[4, 3]. dim 0 of input tensor divide equally by world_size.

```
import torch
import torch.distributed as dist
import os

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

def cleanup():
    dist.destroy_process_group()

def test_nccl_all_to_all(rank, world_size):
    setup(rank, world_size)
    try:
        # shape 0 must be divisible by world_size, other have no requirement
        input_t = (torch.randn((world_size, 4)) ).to(device=f'cuda:{rank}', dtype=torch.bfloat16)
        if rank == world_size - 1:
            input_t = (torch.randn((world_size, 3))).to(device=f'cuda:{rank}', dtype=torch.bfloat16)
        print(f"rank: {rank} input: ", input_t)   
        torch.cuda.synchronize()

        output = torch.empty_like(input_t)
        dist.all_to_all_single(output, input_t)
        print(f"rank: {rank} output: ", output)   
    finally:
        cleanup()

def run_nccl_test():
    world_size = torch.cuda.device_count()
    if world_size < 2:
        print("Need at least 2 GPUs to test NCCL all_to_all")
        return
    
    torch.multiprocessing.spawn(
        test_nccl_all_to_all,
        args=(world_size,),
        nprocs=world_size,
        join=True
    )

if __name__ == "__main__":
    run_nccl_test()
```

rank0:output [2, :] received four data from rank3:input, which comes from **rank3:input[0, :] + rank3:input[1, 0]**;

rank1:output [2, 0:2] received rank3:input[1, :] + **rank3:input[1, 3]**;

I wonder if this is a reasonable behavior? Or does all_to_all_sigal have to ensure that all rank inputs have the same shape?
```
rank: 0 input:  tensor([[ 0.9023, -1.0078, -0.0840, -0.5977],
        [ 0.2002, -1.3750, -0.5117, -0.7695],
        [ 1.1562,  0.0037,  1.0000, -0.2676],
        [ 0.4941,  1.2422,  1.2344,  0.2383]], device='cuda:0',
       dtype=torch.bfloat16)
rank: 0 output:  tensor([[ 0.9023, -1.0078, -0.0840, -0.5977],
        [-1.2969,  3.0156,  1.5469,  1.1250],
        [-1.1797, -0.1357, -0.2305,  0.4180],
        [-1.7109,  1.3047,  0.4883, -2.0938]], device='cuda:0',
       dtype=torch.bfloat16)

rank: 1 input:  tensor([[-1.2969,  3.0156,  1.5469,  1.1250],
        [-0.1973, -0.7539, -0.3926,  1.1953],
        [-0.1377, -1.1641,  0.6719, -0.3418],
        [ 0.9141,  0.6094, -0.1934, -0.0132]], device='cuda:1',
       dtype=torch.bfloat16)
rank: 1 output:  tensor([[ 0.2002, -1.3750, -0.5117, -0.7695],
        [-0.1973, -0.7539, -0.3926,  1.1953],
        [-2.2500, -1.5703,  1.1953,  1.2031],
        [-2.0938, -0.2559,  1.5938,  0.0000]], device='cuda:1',
       dtype=torch.bfloat16)

rank: 2 input:  tensor([[-1.1797, -0.1357, -0.2305,  0.4180],
        [-2.2500, -1.5703,  1.1953,  1.2031],
        [ 0.5703,  0.2002,  0.1602, -0.8359],
        [-0.0356, -0.4473,  1.0234,  1.7422]], device='cuda:2',
       dtype=torch.bfloat16)
rank: 2 output:  tensor([[ 1.1562,  0.0037,  1.0000, -0.2676],
        [-0.1377, -1.1641,  0.6719, -0.3418],
        [ 0.5703,  0.2002,  0.1602, -0.8359],
        [ 0.0337,  0.6836,  0.6172,  1.3594]], device='cuda:2',
       dtype=torch.bfloat16)

rank: 3 input:  tensor([[-1.7109,  1.3047,  0.4883],
        [-2.0938, -0.2559,  1.5938],
        [ 0.0337,  0.6836,  0.6172],
        [ 1.3594,  0.7109, -0.7109]], device='cuda:3', dtype=torch.bfloat16)
rank: 3 output:  tensor([[ 0.4941,  1.2422,  1.2344],
        [ 0.9141,  0.6094, -0.1934],
        [-0.0356, -0.4473,  1.0234],
        [ 1.3594,  0.7109, -0.7109]], device='cuda:3', dtype=torch.bfloat16)
```

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.29.0
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.10.134-13.an8.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe

Nvidia driver version: 535.161.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.1.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          112
On-line CPU(s) list:             0-111
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              28
Socket(s):                       2
Stepping:                        6
CPU max MHz:                     3100.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       2.6 MiB (56 instances)
L1i cache:                       1.8 MiB (56 instances)
L2 cache:                        70 MiB (56 instances)
L3 cache:                        84 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-27,56-83
NUMA node1 CPU(s):               28-55,84-111
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.0
[pip3] torchvision==0.22.0
[pip3] triton==3.3.0
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] torch                     2.7.0                    pypi_0    pypi
[conda] torchvision               0.22.0                   pypi_0    pypi
[conda] triton                    3.3.0                    pypi_0    pypi
